File: LeetCode/0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int len = nums.length;
        int product = 1;
        int[] answer = new int[len];
        
        for (int i = 0; i < len; i++) {
            answer[i] = product;
            product *= nums[i];
        }
        
        product = 1;
        
        for (int i = len - 1; i >= 0; i--) {
            answer[i] *= product;
            product *= nums[i];
        }
        
        return answer;
    }
}

File: LeetCode/0238-product-of-array-except-self/0238-product-of-array-except-self.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        var p = 1
        val answer = IntArray(nums.size)
        
        for (i in nums.indices) {
            answer[i] = p
            p *= nums[i]
        }
        
        p = 1
        
        for (i in nums.indices.reversed()) {
            answer[i] *= p
            p *= nums[i]
        }
        
        return answer
    }
}